aten/src/ATen/Parallel.h

```diff
@@ -32,6 +33,19 @@ CAFFE2_API int get_thread_num();
 // Checks whether the code runs in parallel region
 CAFFE2_API bool in_parallel_region();
 
+namespace internal {
+
+// Initialise num_threads lazily at first parallel call
```

Comment:
It would be good to have more explicit instructions about when to use this. AFAICT, you have to call this before every omp pragma?

Reply:
unfortunately there are are some new pragmas that got into the codebase, I specifically killed all of them earlier and used at::parallel_for, but I'll follow up with a diff to clean them up

In general, we do call `at::init_num_threads` in places we control (i.e. except explicit user created threads); this PR is more a safe guard to make sure the original settings are respected if user creates a new thread themselves and user uses at::paralllel_for
So, technically after this PR the only case we don't cover is when user explicitly creates their own thread and user doesn't use at::parallel_for (and e.g. uses OpenMP directly), but in this case we can't really do much anyways